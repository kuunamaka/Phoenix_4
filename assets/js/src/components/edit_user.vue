<template>
  <table>
    <div>
      <h2>Updating User</h2>
      <h4>* User id won't be able to change</h4>
      <p v-if = "errors.length">
        Please correct the following error(s):
        <ul>
          <li v-for = "error in errors" :key="error" class = "error"> {{ error }}</li>
        </ul>
      </p>
      <input v-model="name" placeholder="new name">
      <input v-model="username" placeholder="new username">
    </div>
    <br />
      <button @click="updateUser(user)">Submit</button>
      <a href="/users" class="button">Exit</a>
    <div>
      <delete-user />
    </div>
  </table>
</template>

<script>
import axios from 'axios'
import DeleteUser from './delete_user';
export default {
  name: 'user_detail',
  components: {DeleteUser},
  data() {
    return {
      errors: [],
      name: '',
      username: '',
      user: ''
    }
  },
  methods: {
    async updateUser(user) {
      if (this.name && this.username) {
        await axios.put('/api/users' + `/${ user.id }`, {
          user: {
            name: this.name,
            username: this.username
          }
        })
        window.location.href = '/users';
      }
      
      this.errors = [];

      if (!this.name) {
        this.errors.push("Name required");
      }
      if (!this.username) {
        this.errors.push("Username required");
      }
    }
  },
  async mounted() {
    const user_id = window.location.pathname.split('/')[2];
    const resp = await axios.get('/api' + `/users/${ user_id }`)
    this.user = resp.data
  }
}
</script>
