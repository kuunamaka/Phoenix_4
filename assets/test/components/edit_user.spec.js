import { shallowMount } from '@vue/test-utils'
import edit_user from  '@/src/components/edit_user.vue'

const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const mock = new  MockAdapter(axios);

mock.onGet("/api/users/1").reply(200, {
  id: 1, name: "Test", username: "testtest"}
);

global.window = Object.create(window);
const pathname = '/users/1';

Object.defineProperty(window, 'location', {
  value: {
    pathname,
    href: `http://localhost:4001${ pathname }`
  }
});

describe('edit_user.vue', () => {
  const wrapper = shallowMount(edit_user);
  const vm = wrapper.vm
  
  it('renders entire component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('updateUser()', () => {
    it('redirect to /users', async () => { 
      mock.onPut('/api/users/1').reply(200, {user: {id: 1, name: 'newTest', username: 'newtesttest'}});
      await wrapper.setData({id: 1, name: 'newTest', username: 'newtesttest'})

      await vm.updateUser({id: 1, name: 'newTest', username: 'newtesttest'})

      expect(window.location.href).toEqual('/users');
    });

    it('show validate error messages', async () => {
      mock.onPut("/api/users/1", {user: {id: 1, name: '', username: ''}}).reply(404);
      await wrapper.setData({id: 1, name: '', username: ''});
  
      const updateButton = wrapper.find('button');
      await updateButton.trigger('click');
  
      expect(wrapper.find('p').element).toMatchSnapshot();
    });
  })
})
