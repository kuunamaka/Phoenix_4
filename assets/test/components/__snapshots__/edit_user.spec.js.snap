// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit_user.vue is a Vue instance 1`] = `
<table>
  <div>
    <h2>
      Updating User
    </h2>
     
    <h4>
      * User id won't be able to change
    </h4>
     
    <!---->
     
    <input
      placeholder="new name"
    />
     
    <input
      placeholder="new username"
    />
  </div>
   
  <br />
   
  <button>
    Submit
  </button>
   
  <a
    class="button"
    href="/users"
  >
    Exit
  </a>
   
  <div>
    <delete-user-stub />
  </div>
</table>
`;

exports[`edit_user.vue updateUser() show validate error messages 1`] = `
<table>
  <div>
    <h2>
      Updating User
    </h2>
     
    <h4>
      * User id won't be able to change
    </h4>
     
    <p>
      
      Please correct the following error(s):
      
      <ul>
        <li
          class="error"
        >
           Name required
        </li>
        <li
          class="error"
        >
           Username required
        </li>
      </ul>
    </p>
     
    <input
      placeholder="new name"
    />
     
    <input
      placeholder="new username"
    />
  </div>
   
  <br />
   
  <button>
    Submit
  </button>
   
  <a
    class="button"
    href="/users"
  >
    Exit
  </a>
   
  <div>
    <delete-user-stub />
  </div>
</table>
`;
